lib/core/macros.nim

```diff
@@ -1722,3 +1722,127 @@ proc extractDocCommentsAndRunnables*(n: NimNode): NimNode =
         result.add ni
       else: break
     else: break
+
+macro stamp*(body: untyped): NimNode =
+  ## Accepts a template body and returns the AST that represents it.
+  ##
+  ## Identifiers within `body` are bound to symbols from the caller's scope in
+  ## the same fashion as a template. As a special case, `result` is excluded
+  ## from automatic binding.
+  ##
+  ## The template body is hygienic, as such identifiers declared within might
+  ## turn into `gensym` symbols. This behavior can be overridden using
+  ## `{.gensym.}` or `{.inject.}` pragmas at declaration sites. Consult the
+  ## language manual for details on template hygiene.
+  ##
+  ## Within `body`, placeholders, which are references to `NimNode` values in
+  ## the caller's scope delimited by backticks, are substituted with copies of
+  ## the referenced values.
```

Comment:
This part needs to be adjusted:
* the unquoted expressions can be of any type, not just `NimNode`
* only the reference is copied for `NimNode`s, not the node itself (this is not clear at the moment)